src/Dependencies/PooledObjects/ArrayBuilder.cs

```diff
@@ -431,15 +439,20 @@ public static ArrayBuilder<T> GetInstance()
 
         public static ArrayBuilder<T> GetInstance(int capacity)
         {
-            var builder = GetInstance();
+            var builder = s_emptyBuilder;
+            if (builder == null ||
+                builder != Interlocked.CompareExchange(ref s_emptyBuilder, null, builder))
```

Comment:
>builder != Interlocked.CompareExchange(ref s_emptyBuilder, null, builder) [](http://example.com/codeflow?start=16&length=73)

Attempts to avoid using a builder that has already allocated an underlying array that might need to be thrown away because it's not big enough.